Fix fwrite() truncation on resource

ktomk opened the pull request:
When using fwrite() operation on a resource, it is not a given to expect
that all data has been written in the operation.

This fix still checks for the error return value, but also if all bytes
have been written and if not all bytes have yet been written, the write
operation will continue until all bytes have been written _or_ an error
occurs.

This is particularly important for operation on network streams.

MaxSavich commented:
In case when fwrite will return 0 (lack of disk space for example) and the code will go to infinite loop.

ktomk commented:
@MaxSavich: If the reason that 0 bytes have been written is not an error (in the disk full example you give) the program will continue to write until all bytes have been written. That's most likely not endless because a system can't continue to operate on a full disk system. After space is freed (to stay in your example), the program continues to write the file until done.

If freeing disk space needs system interruption, the program is interrupted with the system as well.

So even for your quite hypothetical example (which is technically correctly constructed, fwrite returns 0 in that case and not FALSE), it is not an infinite loop but a conditional one and the condition is not theoretically infinite but just constructed (no disk space) and only lasts as long as the construction - which is not infinite.

MaxSavich commented:
@ktomk there is one more scenario described here: http://php.net/manual/en/function.fwrite.php#96951
I didn't try to reproduce it and comment is 5 years old, so maybe it is not relevant any more, but it looks like there is more cases when fwrite returns 0. 
Also, there is many scenarios when it is possible to have no disk space for static content but have it for the system (e.g. separate hard drive for static files, NFS, etc.).

ktomk commented:
Hm, I then guess that 0 is an error code for the underlying operation and it should be treated that way. I've taken a peek into streams in PHP, readme says to follow stdio semantics. http://linux.die.net/man/3/fwrite - 0 or lower strlen is with an error. howerver, 0 should be the only one we can not recover from (apart of freeing space which must not happen quickly). So exit on 0 is necessary, throwing an exception that writing failed.

ktomk committed:
Fix fwrite() truncation on resource

When using fwrite() operation on a resource, it is not a given to expect
that all data has been written in the operation.

This fix still checks for the error return value, but also if all bytes
have been written and if not all bytes have yet been written, the write
operation will continue until all bytes have been written *or* an error
occurs.

This is particularly important for operation on network streams.

MaxSavich commented:
@ktomk your pull request has been merged. Thank you!
